src/main/java/echoServer/SocketReaderWriter.java

```diff
@@ -8,22 +8,31 @@
 
 public class SocketReaderWriter implements ReadableWriteable {
     private final Socket socket;
+    private final BufferedReader reader;
+    private final PrintWriter writer;
 
-    public SocketReaderWriter(Socket socket) {
+    public SocketReaderWriter(Socket socket) throws IOException {
         this.socket = socket;
+        this.reader = new BufferedReader(
+                new InputStreamReader(this.socket.getInputStream())
+        );
+        this.writer = new PrintWriter(socket.getOutputStream(), true);
+
     }
 
     @Override
     public String readLine() throws IOException {
-        BufferedReader reader = new BufferedReader(
-                new InputStreamReader(this.socket.getInputStream())
-        );
         return reader.readLine();
     }
 
     @Override
     public void writeLine(String message) throws IOException {
-        PrintWriter writer = new PrintWriter(socket.getOutputStream(), true);
         writer.println(message);
     }
+
+    @Override
+    public void close() throws IOException {
+        reader.close();
+        writer.close();
```

Comment:
It looks like closing the socket will also close its input and output streams, so if I were to continue with calling .close() instead of using a try-with-resources block, I believe I could just call socket.close() here. But based on your other comment, it seems like it's beneficial to use a try-with-resources block for error handling purposes. 

I'm currently stuck on where to put a try-with-resources block since the Autocloseable resources need to be instantiated within the block itself. I was considering putting it in Main.main() and specifying the ServerSocket as the resource, but I think the Socket needs to be specified as a resource too, and it's created within PortListener.listen().

Reply:
I think you are right that we only need to call `close` on the Socket, and I think it is also true that `ServerSocket` is probably something that needs to be closed since it is a resource.

Regarding the `try-with-resources`, I think we would only use that with objects that we own. For example, we don't need to use `try-with-resources` on the `Socket` or `ServerSocket` because we are wrapping those behind objects that you own.

Reply:
Also, `try-with-resources` isn't required. The message I am trying to send is this

1. We should close the resources that we open.
2. We should ensure that those resources are closed, even in the event of something exceptional.

Reply:
>  I was considering putting it in Main.main() and specifying the ServerSocket as the resource

I think that would work.

> I think the Socket needs to be specified as a resource too

I think you are right. Main won't work here. I think we would need to use `try-with-resources` where we call `listen`, and we would need to make the `ReadableWritable` that we get back something that can be autoclosed.